metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala

```diff
@@ -72,6 +97,24 @@ object ConfDecoder {
         )
     }
 
+  def instanceExpectWithDefault[T](
+      expect: String
+  )(rwd: (Conf, T, Conf => Configured[T]) => Configured[T])(
+      f: PartialFunction[Conf, Configured[T]]
+  )(implicit ev: ClassTag[T]): ConfDecoderWithDefault[T] =
+    new ConfDecoderWithDefault[T] {
+      override def read(any: Conf): Configured[T] =
```

Comment:
i already commented that I'd consider changing this method in `ConfDecoder` to add the `Option [T] = None` if possible

Reply:
The main problem with it is Decoder modification. `ConfDecoder` is invariant by its parameter, so you can't define `map` method without `default` [modification](https://github.com/olafurpg/metaconfig/pull/88/files#diff-8844e7bc4f6dd07a798e11914a4e4baaR48-R63). 

But it's possible to create something like it:

```scala
trait ConfDecoder[T] {
  def read(conf: Conf, default: Option[T] = None): Configured[T]
  final def map[B](f: A => B): ConfDecoder[B] = self.flatMap(x => Ok(f(x)))(g)
  final def flatMap[TT](f: A => Configured[TT]): ConfDecoder[TT] =
    new ConfDecoder[TT] {
      override def read(any: Conf, default: Option[TT]): Configured[TT] =
        self.read(any, None) match { // pass None to default
          case Ok(x) => f(x)
          case NotOk(x) => Configured.NotOk(x)
        }
    }
  final def mapWithDefault[B](f: A => B)(g: B => A): ConfDecoder[B] =
    self.flatMapWithDefault(x => Ok(f(x)))(g)
  final def flatMapWithDefault[TT](f: A => Configured[TT])(g: TT => A): ConfDecoder[TT] =
    new ConfDecoder[TT] {
      override def read(any: Conf, default: Option[TT]): Configured[TT] =
        self.read(any, default.map(g)) match { // be honest with default
          case Ok(x) => f(x)
          case NotOk(x) => Configured.NotOk(x)
        }
    }
```

Disadvantage: if user calls `.map` on `Decoder` for list he will miss default value without a hint from the types.

Advantage: simplicity in contrast with current implementation.

Reply:
> The main problem with it is Decoder modification. `ConfDecoder` is invariant by its parameter, so you can't define `map` method without `default` [modification]

next question: is it possible to modify the current `generic.deriveDecoder(...)` to turn it into a factory which takes a `default: T`? i think the difficulty is we have a decoder which was created with an initial instance and then we want to use a different instance; having a factory would solve this problem.

i don't know how macros work, and whether `Macros.deriveConfDecoderImpl` can be converted into a factory possibly explicitly passing a weak type tag to it to instantiate it.

what do you think?

Reply:
See #90

Reply:
The factory approach looks interesting. Combined with presets I feel like we may be able to achieve an OK solution for Scalafmt